Framework: Removed lib/site/computed-attributes dependency from state selectors.

jorgefilipecosta opened the pull request:
This PR removes lib/site/computed-attributes dependency from state/sites/selectors.js.

This PR is dependent on pull request https://github.com/Automattic/wp-calypso/pull/17154 that updates test cases to take into consideration that getSite selector requires user information in the state.

The getSiteComputedAttributes uses the canCurrentUser selector because is the recommended way to check for user capabilities.
This dependency has a big impact. 

The automated tests for sites needed to pass some user state which causes some tests to fail to overcome that the tests were updated to pass user state.

Another implication of the dependency from canCurrentUser is that right now getSite and every other selector that returns an object from getSite (e.g: getJetpackSite, getPlublicSites etc...) depends on state.currentUser.capabilities  part of the state and not just state.sites.items as before. So for the memorized selectors, it was needed to add this dependency.

**To test:**
Use the application giving more relevance to areas that make the most use of getSite selector (plugins/ posts/ settings) and check if no error or strange behaviour happens.

aduth commented on client/state/current-user/selectors.js:
```diff
@@ -104,7 +104,7 @@ export const getCurrentUserEmail = createCurrentUserSelector( 'email' );
  */
 export function isValidCapability( state, siteId, capability ) {
-	const capabilities = state.currentUser.capabilities[ siteId ];
+	const capabilities = get( state, [ 'currentUser', 'capabilities', siteId ] );
```
Why do we want to use `get` here? We know the path to `state.currentUser.capabilities` will always exist because this is the shape of the store reducer. In this case, `get` will needlessly check for truthiness of each property in the complete path.

jorgefilipecosta replied:
The reasons this was used was to not require to going to all automated tests and adding some user state.  I give more details about this in the description.

aduth commented on client/state/sites/selectors.js:
```diff
@@ -86,14 +85,12 @@ export const getSite = createSelector(
 		return site;
 	},
-	( state ) => state.sites.items
+	( state, siteId ) => [ get( state, [ 'sites', 'items', siteId ] ), get( state, [ 'currentUser', 'capabilities', siteId ] ) ]
```
Similar to above, if we never have the risk of the path having an undefined value (we know this to be the case because of the shape of the reducer), we ought to not use `get`.

jorgefilipecosta replied:
Our current test cases do not pass state information for functions that now require it. If get was not used, 44 test cases failed. To make them pass is easy we just need to pass the userState but that would make this PR very very huge. If we decide to update the test cases in order to not use this get. I can do some pull requests that update them. And than change this one.

jorgefilipecosta commented:
Hi @aduth , thank you a lot for landing the dependency of this PR. Regarding this one I would like to know, what is your option we should change 40+ test cases to add user state them or we can use get in this case to avoid the huge change in the test cases?
(PR was rebased)

aduth commented:
>Regarding this one I would like to know, what is your option we should change 40+ test cases to add user state them or we can use get in this case to avoid the huge change in the test cases?

I don't know that I have a good answer. I'd say ideally we wouldn't use `get`, specifically because we know the shape of the reducer will guarantee those properties exist in state, in which case we're allowing our implementation to be dictated by the needs of testing (usually a bad thing).

That said, this is a common pain point for testing selectors, and needs to be better addressed at a framework level. A few ideas that come to mind off the top of my head. Let me know if you have any thoughts.

**Create a state object by merging the test values into the actual reducer:**

```js
import { merge } from 'lodash';
import { reducer } from 'state';
import { hasThing } from 'state/selectors';

describe( 'hasThing()', () => {
	it( 'should return true if thing exists', () => {
		const state = merge(
			reducer( undefined, {} ),
			{
				things: {
					exists: true
				}
			}
		);

		expect( hasThing( state ) ).to.be.true;
	} );
} );
```

**Mock dependant selectors:**

This one could be done currently with [Mockery](https://github.com/Automattic/wp-calypso/tree/master/test/test/helpers/use-mockery), but will become a little nicer if/when we switch to [Jest](https://facebook.github.io/jest/) (#16119). You can see an example of this in practice in [the Gutenberg project](https://github.com/WordPress/gutenberg/blob/3543d4c/editor/test/effects.js#L159-L166).
 
```js
import * as selectors from 'state/selectors';

jest.mock( 'state/selectors' );

describe( 'hasThingAndIsLarge()', () => {
	beforeEach( () => jest.resetAllMocks() );

	it( 'should return true if thing exists and is large', () => {
		selectors.hasThing.mockReturnValue( true );
		selectors.isThingLarge.mockReturnValue( true );

		expect( hasThingAndIsLarge( state ) ).to.be.true;
	} );
} );
```

The downside to this one is that it approaches testing the implementation moreso than testing the behavior of the function.

jorgefilipecosta commented:
Hi @aduth, thank you for your review and for the suggestions.  I will keep your suggestions in mind next time I have a problem an identical problem in selectors testing.
After rebasing this branch and changing the code to not use get to get properties we know should exist in the state, to my surprise, the number of tests failing was just 8. The last time I checked they were many (40+) I don't know exactly the reason for this reduction maybe something was improved in the tests and rebasing solved that, or the last time I executed the tests before using the "get" I had a problem causing tests to fail that was meanwhile solved.
Given the small number of tests failing and given that I had a user state fixture, I just passed the userState to the tests that were failing because of it.

Right now automated tests are failing in CI because the dependencies are not merged yet. I locally merged them and tests are ok.

I think all your feedback was addressed feel free to have a new look.

aduth commented:
Can you rebase this given merge of #16938?

jorgefilipecosta commented:
Hi @aduth,
This PR was rebased. Meanwhile, I found some tests of selectors outside selectors folder that also needed user state I added it to them and now we have a green in CircleCI.
But this PR got a little hard to review, I'm thinking about separating test changes to another PR to make things easier. 
I will also need to verify the selectors whose tests I changed in more detail and check if everything looks fine. So I will temporarily put this as in progress. When I have things ready for a new look I will comment. Thank you for the help @aduth.

jorgefilipecosta commented:
Hi @aduth ,
I rebased this PR and I separated in another pull request just for the addition of userState/capabilities to test cases https://github.com/Automattic/wp-calypso/pull/17154  to make things easier to review.
I did a new review and I found a condition that needed to be fixed in getSiteComputedAttributes, I also added some comments there (separated commit).

Thank you for your efforts in reviewing this changes!

aduth commented:
There are some failing tests in the current branch.

jorgefilipecosta commented:
Hi @aduth, the reason of the failed tests is that now I separated test changes to another PR https://github.com/Automattic/wp-calypso/pull/17154. That PR just adds user information to test cases,  nothing more. As soon as that one is landed tests on this one are green. We can test it by merging the other branch in this one.

jorgefilipecosta commented:
There was a very recent change to getSitesWithoutPlugin selector and now it calls the getSite selector so it's tests need userState. I added the userState to this selector in this PR instead of in PR https://github.com/Automattic/wp-calypso/pull/17154 because at the time of that PR userState was not needed. Right now automatic tests are ok. I will do further manual testing to make sure this is good.

gwwar commented:
Since this might affect a lot of the app, let's also run the e2e suite against the branch. @hoverduck would that be possible?

hoverduck commented:
I ran the e2e suite against both wpcom and Jetpack sites and didn't see any real issues.  I had problems with the Customizer tests, but I think that was just due to a slow internet connection causing timeouts.  Manual checkout of the customizer was fine.

gwwar commented:
Thanks @hoverduck !

jorgefilipecosta commented:
Hi @aduth, @gwwar and @hoverduck thank you for your reviews and testing. I also did a new set of tests and compared the old logic with the new one and things seem fine.

I think now I'm able to deploy so I will get in touch with someone who is available to help me with that. Probably I will try to deploy something less risky before deploying this one.

tyxla commented:
Thank you @jorgefilipecosta! 👍
